src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -177,6 +177,21 @@ private bool CompleteCore()
             {
                 if (existingWaiters != null)
                 {
+                    // we have waiters and we have data.
+                    // allow some extra time to drain the queue before cancelling
+                    // but do not wait for too long.
+                    SpinWait.SpinUntil(() =>
+                    {
+                        lock (SyncObject)
+                        {
+                            return _data.Count == 0;
+                        }
+                    },
+                    10000);
```

Comment:
Then you should leave a comment and make a constant or, preferably, use `TimeSpan.FromSeconds(10)` to make this more readable. Without looking up SpinUntil I don't know that's a timeout or what units it's in.